File: lib/src/book.dart
part of 'libxl_ext.dart';

class XlBook {
  final int ptr;

  XlBook() : ptr = _xlCreateBook();

  const XlBook.empty() : ptr = 0;

  @override
  String toString() {
    return 'XlBook{ptr: $ptr}';
  }

  void release() {
    _xlBookRelease(ptr);
  }

  void save(String fileName) {
    int res = _xlBookSave(ptr, fileName);
    if (res == 0) {
      throw XlException(errorMessage());
    }
  }

  String errorMessage() {
    return _xlBookErrorMessage(ptr);
  }

  XlSheet getSheet(int index) {
    int sheetPointer = _xlBookGetSheet(ptr, index);
    if (sheetPointer == 0) {
      throw XlException(errorMessage());
    }
    return XlSheet(sheetPointer, this);
  }

  XlSheet addSheet(String sheetName,
      [XlSheet initSheet = const XlSheet.empty()]) {
    int sheetPointer = _xlBookAddSheet(ptr, sheetName, initSheet.ptr);
    if (sheetPointer == 0) {
      throw XlException(errorMessage());
    }
    return XlSheet(sheetPointer, this);
  }

  XlFormat addFormat({XlFormat initFormat = const XlFormat.empty()}){
    var f = _xlBookAddFormat(ptr, initFormat.ptr);
    var format = new XlFormat(f, this);
    return format;
  }
}

File: lib/src/c/book.cc
void _xlCreateBook(Dart_NativeArguments args) {
  Dart_EnterScope();
  BookHandle book = xlCreateBook();
  Dart_Handle result = Dart_NewInteger((int64_t)book);
  Dart_SetReturnValue(args, result);
  Dart_ExitScope();
}
void _xlBookSave(Dart_NativeArguments args) {
  Dart_EnterScope();
  int64_t ptr;
  Dart_IntegerToInt64(Dart_GetNativeArgument(args, 0), &ptr);
  const char* fileName;
  Dart_StringToCString(Dart_GetNativeArgument(args, 1), &fileName);
//  Book *book = reinterpret_cast<Book*>(ptr);
//  bool res = book->save(fileName);
  int res = xlBookSave((BookHandle) ptr, fileName);
  Dart_Handle result = Dart_NewBoolean(res != 0);
  Dart_SetReturnValue(args, result);
  Dart_ExitScope();
}
void _xlBookRelease(Dart_NativeArguments args) {
  Dart_EnterScope();
  int64_t ptr;
  Dart_IntegerToInt64(Dart_GetNativeArgument(args, 0), &ptr);
  xlBookRelease((BookHandle) ptr);
  Dart_SetReturnValue(args, Dart_Null());
  Dart_ExitScope();
}
void _xlBookAddSheet(Dart_NativeArguments args) {
  Dart_EnterScope();
  int64_t ptr;
  Dart_IntegerToInt64(Dart_GetNativeArgument(args, 0), &ptr);
  const char* sheetName;
  Dart_StringToCString(Dart_GetNativeArgument(args, 1), &sheetName);
  int64_t initSheet;
  Dart_IntegerToInt64(Dart_GetNativeArgument(args, 2), &initSheet);
  SheetHandle sheet = xlBookAddSheet((BookHandle) ptr, sheetName, (SheetHandle) initSheet);
  Dart_Handle result = Dart_NewInteger((int64_t)sheet);
  Dart_SetReturnValue(args, result);
  Dart_ExitScope();
}
void _xlBookGetSheet(Dart_NativeArguments args) {
  Dart_EnterScope();
  int64_t ptr;
  Dart_IntegerToInt64(Dart_GetNativeArgument(args, 0), &ptr);
  int64_t index;
  Dart_IntegerToInt64(Dart_GetNativeArgument(args, 1), &index);
  SheetHandle sheet = xlBookGetSheet((BookHandle) ptr, index);
  Dart_Handle result = Dart_NewInteger((int64_t)sheet);
  Dart_SetReturnValue(args, result);
  Dart_ExitScope();
}

void _xlBookErrorMessage(Dart_NativeArguments args) {
  Dart_EnterScope();
  int64_t ptr;
  Dart_IntegerToInt64(Dart_GetNativeArgument(args, 0), &ptr);
  const char* errorMessage = xlBookErrorMessage((BookHandle) ptr);
  Dart_Handle result = Dart_NewStringFromCString(errorMessage);
  Dart_SetReturnValue(args, result);
  Dart_ExitScope();
}
